provider/awssd/aws_sd.go

```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
 	return creates, deletes
 }
 
+func (p *AWSSDProvider) handleSRVChanges(namespaces []*sd.NamespaceSummary, creates []*endpoint.Endpoint, deletes []*endpoint.Endpoint) ([]*endpoint.Endpoint, []*endpoint.Endpoint, error) {
+	//create a struct of all relevant changes to handle SRV-related changes
+	type relevantChanges struct {
+		ACre   []*endpoint.Endpoint // max len is 1
+		SRVCre []*endpoint.Endpoint // max len is 1
+		SRVDel []*endpoint.Endpoint // max len is n
+	}
+
+	createsByNamespaceID := p.changesByNamespaceID(namespaces, creates)
+	deletesByNamespaceID := p.changesByNamespaceID(namespaces, deletes)
+
+	for _, ns := range namespaces {
+		nsID := *ns.Id
+		nsCreates, okNsCre := createsByNamespaceID[nsID]
+		nsDeletes, okNsDel := deletesByNamespaceID[nsID]
+		//loop all namespaces where we have at least 1 change
+		if !(okNsCre || okNsDel) {
+			continue
+		}
+
+		services, err := p.ListServicesByNamespaceID(aws.String(nsID))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		//create a map of relevant changes by host. will loop this map later to determine any SRV-related changes
+		changesByHost := make(map[string]*relevantChanges)
+		if okNsCre {
+			for _, ch := range nsCreates {
+				// add a Create record type to the changesByHost. Add max 1 because only 1 SRV is supported for 1 host
+				if ch.RecordType == endpoint.RecordTypeSRV {
+					_, host, _, _, err := p.srvHostTargetSplit(ch.Targets[0])
+					if err != nil {
+						return nil, nil, err
+					}
+					if _, ok := changesByHost[host]; !ok {
+						changesByHost[host] = &relevantChanges{}
+					}
+					chByHost := changesByHost[host]
+					if len(chByHost.SRVCre) == 0 {
+						chByHost.SRVCre = append(chByHost.SRVCre, ch)
+					}
+				}
+				// add a Create record type to the changesByHost. Add max 1 because only 1 A is expected for 1 host
+				if ch.RecordType == endpoint.RecordTypeA {
+					host := ch.DNSName
+					if _, ok := changesByHost[host]; !ok {
+						changesByHost[host] = &relevantChanges{}
+					}
+					chByHost := changesByHost[host]
+					if len(chByHost.ACre) == 0 {
+						chByHost.ACre = append(chByHost.ACre, ch)
+					} else {
+						log.Warnf("Skipping record %s because only 1 create change is expected for A record. This is unexpected.", ch.String())
+					}
```

Comment:
I believe it requires a high level comment describing the data transformation. The code is quite complex.

Reply:
added a higher level description